### Initial data cleanup

In [1]:
import pandas as pd

RAW_CSV_FILEPATH = '../data/raw_scraped_car_listings.csv'
CLEANED_CSV_FILEPATH = '../data/cleaned_car_listings.csv'

In [2]:
df = pd.read_csv(RAW_CSV_FILEPATH)
df

,make,model,year,body_type,mileage,door_count,nr_seats,color,fuel_type,engine_capacity,engine_power,gearbox,transmission,pollution_standard,price,url
0,BMW,X3,2015,SUV,194 382 km,5,5,Maro,Diesel,1 995 cm3,190 CP,Automata,4x4 (automat),Euro 6,15 990,https://www.autovit.ro/autoturisme/anunt/bmw-x...
1,Volvo,XC 60,2015,SUV,122 822 km,5,5,Albastru,Diesel,2 400 cm3,181 CP,Automata,4x4 (automat),Euro 5,14 000,https://www.autovit.ro/autoturisme/anunt/volvo...
2,Mercedes-Benz,B,2005,Monovolum,249 000 km,5,5,Albastru,Diesel,1 991 cm3,140 CP,Manuala,Fata,Euro 4,3 000,https://www.autovit.ro/autoturisme/anunt/merce...
3,Peugeot,508,2012,Sedan,212 000 km,4,5,Negru,Hibrid,1 997 cm3,163 CP,Automata,4x4 (automat),Euro 6,7 400,https://www.autovit.ro/autoturisme/anunt/peuge...
4,Hyundai,Tucson,2021,SUV,44 000 km,5,5,Gri,Hibrid Plug-In,1 598 cm3,265 CP,Automata,4x4 (automat),Euro 6,"26 607,90",https://www.autovit.ro/autoturisme/anunt/hyund...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,BMW,Seria 3,2014,Combi,267 000 km,5,5,Gri,Diesel,1 995 cm3,143 CP,Manuala,Spate,Euro 5,8 498,https://www.autovit.ro/autoturisme/anunt/bmw-s...
1996,Peugeot,5008,2019,SUV,229 000 km,5,5,Alte culori,Diesel,1 499 cm3,130 CP,Automata,Fata,Euro 6,14 700,https://www.autovit.ro/autoturisme/anunt/peuge...
1997,Mercedes-Benz,X,2018,SUV,136 458 km,4,5,Negru,Diesel,2 298 cm3,190 CP,Automata,4x4 (automat),Euro 6,24 400,https://www.autovit.ro/autoturisme/anunt/merce...
1998,Peugeot,308,2018,Combi,137 000 km,5,5,Gri,Benzina,1 199 cm3,130 CP,Manuala,Fata,Euro 6,8 400,https://www.autovit.ro/autoturisme/anunt/peuge...


In [3]:
df['url'].nunique()

2000

The **url** column was used to prevent scraping the same listing page twice, it's not a relevant feature, so we'll discard it.

In [4]:
# drop the url column
df = df.drop(columns=['url'])
df

,make,model,year,body_type,mileage,door_count,nr_seats,color,fuel_type,engine_capacity,engine_power,gearbox,transmission,pollution_standard,price
0,BMW,X3,2015,SUV,194 382 km,5,5,Maro,Diesel,1 995 cm3,190 CP,Automata,4x4 (automat),Euro 6,15 990
1,Volvo,XC 60,2015,SUV,122 822 km,5,5,Albastru,Diesel,2 400 cm3,181 CP,Automata,4x4 (automat),Euro 5,14 000
2,Mercedes-Benz,B,2005,Monovolum,249 000 km,5,5,Albastru,Diesel,1 991 cm3,140 CP,Manuala,Fata,Euro 4,3 000
3,Peugeot,508,2012,Sedan,212 000 km,4,5,Negru,Hibrid,1 997 cm3,163 CP,Automata,4x4 (automat),Euro 6,7 400
4,Hyundai,Tucson,2021,SUV,44 000 km,5,5,Gri,Hibrid Plug-In,1 598 cm3,265 CP,Automata,4x4 (automat),Euro 6,"26 607,90"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,BMW,Seria 3,2014,Combi,267 000 km,5,5,Gri,Diesel,1 995 cm3,143 CP,Manuala,Spate,Euro 5,8 498
1996,Peugeot,5008,2019,SUV,229 000 km,5,5,Alte culori,Diesel,1 499 cm3,130 CP,Automata,Fata,Euro 6,14 700
1997,Mercedes-Benz,X,2018,SUV,136 458 km,4,5,Negru,Diesel,2 298 cm3,190 CP,Automata,4x4 (automat),Euro 6,24 400
1998,Peugeot,308,2018,Combi,137 000 km,5,5,Gri,Benzina,1 199 cm3,130 CP,Manuala,Fata,Euro 6,8 400


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   make                2000 non-null   object
 1   model               2000 non-null   object
 2   year                2000 non-null   int64 
 3   body_type           2000 non-null   object
 4   mileage             2000 non-null   object
 5   door_count          2000 non-null   int64 
 6   nr_seats            2000 non-null   int64 
 7   color               2000 non-null   object
 8   fuel_type           2000 non-null   object
 9   engine_capacity     2000 non-null   object
 10  engine_power        2000 non-null   object
 11  gearbox             2000 non-null   object
 12  transmission        2000 non-null   object
 13  pollution_standard  2000 non-null   object
 14  price               2000 non-null   object
dtypes: int64(3), object(12)
memory usage: 234.5+ KB


In the data collection process, the listings that did not contain values for the selected features were ignored, so we don't have any missing values in the dataset.

In [6]:
# display missing values count
df.isnull().sum()

make                  0
model                 0
year                  0
body_type             0
mileage               0
door_count            0
nr_seats              0
color                 0
fuel_type             0
engine_capacity       0
engine_power          0
gearbox               0
transmission          0
pollution_standard    0
price                 0
dtype: int64

#### Data cleanup

For the **categorical columns** cleanup the strings (remove trailing whitespaces):

In [7]:
categorical_columns = [
    "make",
    "model",
    "body_type",
    "color",
    "fuel_type",
    "gearbox",
    "transmission",
    "pollution_standard"
]

In [8]:
def cleanup_string(string):
    # remove trailing whitespaces
    string = ' '.join(string.split())

    return string

for col in categorical_columns:
    df[col] = df[col].apply(cleanup_string)

df

,make,model,year,body_type,mileage,door_count,nr_seats,color,fuel_type,engine_capacity,engine_power,gearbox,transmission,pollution_standard,price
0,BMW,X3,2015,SUV,194 382 km,5,5,Maro,Diesel,1 995 cm3,190 CP,Automata,4x4 (automat),Euro 6,15 990
1,Volvo,XC 60,2015,SUV,122 822 km,5,5,Albastru,Diesel,2 400 cm3,181 CP,Automata,4x4 (automat),Euro 5,14 000
2,Mercedes-Benz,B,2005,Monovolum,249 000 km,5,5,Albastru,Diesel,1 991 cm3,140 CP,Manuala,Fata,Euro 4,3 000
3,Peugeot,508,2012,Sedan,212 000 km,4,5,Negru,Hibrid,1 997 cm3,163 CP,Automata,4x4 (automat),Euro 6,7 400
4,Hyundai,Tucson,2021,SUV,44 000 km,5,5,Gri,Hibrid Plug-In,1 598 cm3,265 CP,Automata,4x4 (automat),Euro 6,"26 607,90"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,BMW,Seria 3,2014,Combi,267 000 km,5,5,Gri,Diesel,1 995 cm3,143 CP,Manuala,Spate,Euro 5,8 498
1996,Peugeot,5008,2019,SUV,229 000 km,5,5,Alte culori,Diesel,1 499 cm3,130 CP,Automata,Fata,Euro 6,14 700
1997,Mercedes-Benz,X,2018,SUV,136 458 km,4,5,Negru,Diesel,2 298 cm3,190 CP,Automata,4x4 (automat),Euro 6,24 400
1998,Peugeot,308,2018,Combi,137 000 km,5,5,Gri,Benzina,1 199 cm3,130 CP,Manuala,Fata,Euro 6,8 400


For all numerical columns (year, door_count, nr_seats, engine_capacity, engine_power, price), remove whitespace, other string formats (km, cm3, CP) and convert to float type.

For `mileage`, convert from string format "194 382 km" to numerical value, remove "km".

For `engine_capacity`, convert from string format "1 995 cm3" to numerical value, remove "cm3".

For `engine_power`, convert from string format "190 CP" to numerical value, remove "CP".

For `price`, convert commas ',' to decimal points '.'.

In [9]:
# check that all entries have a value in the mileage column that matches the pattern "number km"
matches = df['mileage'].str.match(r'^\d+(\s?\d+)*\s+km$')
print(f'[{matches.sum()}]', "All values from the mileage column respect the format <number km>"if matches.sum() == len(df) else "Unexpected format in mileage column")

# check that all entries have a value in the engine_capacity column that matches the pattern "number cm3"
matches = df['engine_capacity'].str.match(r'^\d+(\s?\d+)*\s+cm3$')
print(f'[{matches.sum()}]', "All values from the engine_capacity column respect the format <number cm3>"if matches.sum() == len(df) else "Unexpected format in engine_capacity column")

# check that all entries have a value in the engine_power column that matches the pattern "number CP"
matches = df['engine_power'].str.match(r'^\d+(\s?\d+)*\s+CP$')
print(f'[{matches.sum()}]', "All values from the engine_power column respect the format <number CP>"if matches.sum() == len(df) else "Unexpected format in engine_power column")

[2000] All values from the mileage column respect the format <number km>
[2000] All values from the engine_capacity column respect the format <number cm3>
[2000] All values from the engine_power column respect the format <number CP>


In [10]:
df['mileage'] = df['mileage'].str.replace('km', '')
df['engine_capacity'] = df['engine_capacity'].str.replace('cm3', '')
df['engine_power'] = df['engine_power'].str.replace('CP', '')
df['price'] = df['price'].str.replace(',','.')
df

,make,model,year,body_type,mileage,door_count,nr_seats,color,fuel_type,engine_capacity,engine_power,gearbox,transmission,pollution_standard,price
0,BMW,X3,2015,SUV,194 382,5,5,Maro,Diesel,1 995,190,Automata,4x4 (automat),Euro 6,15 990
1,Volvo,XC 60,2015,SUV,122 822,5,5,Albastru,Diesel,2 400,181,Automata,4x4 (automat),Euro 5,14 000
2,Mercedes-Benz,B,2005,Monovolum,249 000,5,5,Albastru,Diesel,1 991,140,Manuala,Fata,Euro 4,3 000
3,Peugeot,508,2012,Sedan,212 000,4,5,Negru,Hibrid,1 997,163,Automata,4x4 (automat),Euro 6,7 400
4,Hyundai,Tucson,2021,SUV,44 000,5,5,Gri,Hibrid Plug-In,1 598,265,Automata,4x4 (automat),Euro 6,26 607.90
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,BMW,Seria 3,2014,Combi,267 000,5,5,Gri,Diesel,1 995,143,Manuala,Spate,Euro 5,8 498
1996,Peugeot,5008,2019,SUV,229 000,5,5,Alte culori,Diesel,1 499,130,Automata,Fata,Euro 6,14 700
1997,Mercedes-Benz,X,2018,SUV,136 458,4,5,Negru,Diesel,2 298,190,Automata,4x4 (automat),Euro 6,24 400
1998,Peugeot,308,2018,Combi,137 000,5,5,Gri,Benzina,1 199,130,Manuala,Fata,Euro 6,8 400


For all numerical columns, remove whitespace and convert to float type.

In [11]:
numerical_columns = [
    "year",
    "mileage",
    "door_count",
    "nr_seats",
    "engine_capacity",
    "engine_power",
    "price"
]

for col in numerical_columns:
    if df[col].dtype == 'object':
        df[col] = df[col].str.replace(' ', '')
    df[col] = df[col].astype(float)

df

,make,model,year,body_type,mileage,door_count,nr_seats,color,fuel_type,engine_capacity,engine_power,gearbox,transmission,pollution_standard,price
0,BMW,X3,2015.0,SUV,194382.0,5.0,5.0,Maro,Diesel,1995.0,190.0,Automata,4x4 (automat),Euro 6,15990.0
1,Volvo,XC 60,2015.0,SUV,122822.0,5.0,5.0,Albastru,Diesel,2400.0,181.0,Automata,4x4 (automat),Euro 5,14000.0
2,Mercedes-Benz,B,2005.0,Monovolum,249000.0,5.0,5.0,Albastru,Diesel,1991.0,140.0,Manuala,Fata,Euro 4,3000.0
3,Peugeot,508,2012.0,Sedan,212000.0,4.0,5.0,Negru,Hibrid,1997.0,163.0,Automata,4x4 (automat),Euro 6,7400.0
4,Hyundai,Tucson,2021.0,SUV,44000.0,5.0,5.0,Gri,Hibrid Plug-In,1598.0,265.0,Automata,4x4 (automat),Euro 6,26607.9
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1995,BMW,Seria 3,2014.0,Combi,267000.0,5.0,5.0,Gri,Diesel,1995.0,143.0,Manuala,Spate,Euro 5,8498.0
1996,Peugeot,5008,2019.0,SUV,229000.0,5.0,5.0,Alte culori,Diesel,1499.0,130.0,Automata,Fata,Euro 6,14700.0
1997,Mercedes-Benz,X,2018.0,SUV,136458.0,4.0,5.0,Negru,Diesel,2298.0,190.0,Automata,4x4 (automat),Euro 6,24400.0
1998,Peugeot,308,2018.0,Combi,137000.0,5.0,5.0,Gri,Benzina,1199.0,130.0,Manuala,Fata,Euro 6,8400.0


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   make                2000 non-null   object 
 1   model               2000 non-null   object 
 2   year                2000 non-null   float64
 3   body_type           2000 non-null   object 
 4   mileage             2000 non-null   float64
 5   door_count          2000 non-null   float64
 6   nr_seats            2000 non-null   float64
 7   color               2000 non-null   object 
 8   fuel_type           2000 non-null   object 
 9   engine_capacity     2000 non-null   float64
 10  engine_power        2000 non-null   float64
 11  gearbox             2000 non-null   object 
 12  transmission        2000 non-null   object 
 13  pollution_standard  2000 non-null   object 
 14  price               2000 non-null   float64
dtypes: float64(7), object(8)
memory usage: 234.5+ KB


In [13]:
# find duplicates
df.duplicated().sum()

0

Save the cleaned data

In [14]:
df.to_csv(CLEANED_CSV_FILEPATH, index=False)